File: perf/simple/simple.js
import fs from "node:fs";

const filepath = "data.txt";
const topN = 30;

try {
  const data = fs.readFileSync(filepath, "utf8");
  const words = data.split(" ");

  const wordMap = new Map();

  // get counts for each word
  words.forEach((word) => {
    const mappedWord = wordMap.get(word);

    if (!mappedWord) {
      wordMap.set(word, 1);
      return;
    }

    wordMap.set(word, mappedWord + 1);
  });

  const sorted = Array.from(wordMap).sort((a, b) => {
    const aCount = a[1];
    const bCount = b[1];
    return bCount - aCount;
  });
  const topWords = sorted.slice(0, topN);

  topWords.forEach(([word, count]) => {
    console.log(`${word}: ${count}`);
  });
} catch (err) {
  console.error(err);
}

File: perf/simple/simple.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

var filename string = "../data.txt"
var topN int8 = 30

type word struct {
	Word  string
	Count int
}

func main() {
	data, err := os.ReadFile(filename)

	if err != nil {
		panic(err)
	}

	text := string(data)
	words := strings.Split(text, " ")
	wordMap := make(map[string]int)

	// Count occurances of each word
	for _, word := range words {
		_, exists := wordMap[word]

		if exists {
			wordMap[word] += 1
		} else {
			wordMap[word] = 1
		}
	}

	// sort the words by count descending
	var sorted []word
	for key, value := range wordMap {
		sorted = append(sorted, word{key, value})
	}

	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Count > sorted[j].Count
	})

	for i := range topN {
		word := sorted[i]
		fmt.Printf("%v:%d\n", word.Word, word.Count)
	}
}
